pkg/operator/controller/dnsnameresolver/resolver.go

```diff
@@ -0,0 +1,340 @@
+		if !exists || resolvedName.minNextLookupTime.Before(minNextLookupTime) {
+			exists = true
+			minNextLookupTime = resolvedName.minNextLookupTime
+			dns = dnsName
+			numIPs = resolvedName.numIPs
+		}
+	}
+	return dns, minNextLookupTime, numIPs, exists
+}
+
+// lookupDNSName sends a DNS lookup request to CoreDNS pod(s). The DNS lookup is performed to
+// trigger an update, if required, of the DNSNameResolver resources matching the DNS name.
+func (res *Resolver) lookupDNSName(dnsName string, numIPs int) error {
+	// By default, the DNS lookup request will be sent to maxCoreDNSPodIPs number of CoreDNS
+	// pods.
+	numCoreDNSPodIPs := maxCoreDNSPodIPs
+	// If the DNS name has 0 or 1 associated IP addresses, then the DNS lookup request will
+	// be sent to only 1 CoreDNS pod.
+	if numIPs <= 1 {
+		numCoreDNSPodIPs = 1
+	}
+
+	// Get the randomly chosen CoreDNS pod IPs.
+	coreDNSPodIPs, err := res.getRandomCoreDNSPodIP(numCoreDNSPodIPs)
+	if err != nil {
+		return err
+	}
+
+	// Send the DNS lookup request to the CoreDNS pods for both A and AAAA type DNS records.
+	for _, recordType := range []uint16{dns.TypeA, dns.TypeAAAA} {
+		for _, coreDNSPodIP := range coreDNSPodIPs {
+			dnsMsg := &dns.Msg{}
+			dnsMsg.SetQuestion(dns.Fqdn(dnsName), recordType)
+			serverStr := net.JoinHostPort(coreDNSPodIP, defaultPort)
+			dnsClient := &dns.Client{
+				Timeout: defaultMinTTL,
+			}
+			if _, _, err := dnsClient.Exchange(dnsMsg, serverStr); err != nil {
+				logrus.Warningf("Failed to lookup DNS name: %s from CoreDNS pod with IP: %s, err: %s", dnsName, coreDNSPodIP, err)
+			}
+		}
+	}
+	return nil
+}
+
+// getRandomCoreDNSPodIP returns randomly chosen CoreDNS pod IPs. The input maxIPs defines
+// the upper limit for the number of CoreDNS pod IPs returned.
+func (res *Resolver) getRandomCoreDNSPodIP(maxIPs int) ([]string, error) {
+	// List all the CoreDNS pod endpointslices.
+	epList := &discoveryv1.EndpointSliceList{}
+	if err := res.corednsEndpointSliceCache.List(context.Background(), epList, &client.ListOptions{}); err != nil {
+		return nil, err
+	}
+
+	// Get all the CoreDNS pod IPs from the list.
+	var ips []string
+	for _, epSlice := range epList.Items {
+		for _, ep := range epSlice.Endpoints {
+			if ep.Conditions.Ready != nil && !*ep.Conditions.Ready {
+				continue
+			}
+			ips = append(ips, ep.Addresses...)
+		}
+	}
+
+	// If no IP is found the return an error.
+	if len(ips) == 0 {
+		return nil, fmt.Errorf("no ips found for the coredns pods")
+	}
+
+	var randomIPs []string
+	r1 := rand.New(rand.NewSource(time.Now().UnixNano()))
+
+	if len(ips) <= maxIPs {
+		// If the number of CoreDNS pod IPs is less than equal to maxIPs,
+		// then return all the CoreDNS pod IPs.
+		randomIPs = ips
+	} else {
+		// If the number of CoreDNS pod IPs is greater than maxIPs, then
+		// return randomly chosen maxIPs number of CoreDNS pod IPs
+		for i := 0; i < maxIPs; i++ {
+			randomIPs = append(randomIPs, ips[r1.Intn(len(ips))])
```

Comment:
The random number chosen for an index could be duplicated, possibly resulting in a smaller set of randomIPs than what you would get if you just took the first 5 ips.

Reply:
I'll update the code to avoid this scenario. The randomness was included so that the DNS queries does not always end up with the same CoreDNS pods.

Reply:
Moved to openshift/coredns-ocp-dnsnameresolver#2